airbyte-integrations/connectors/source-google-drive/source_google_drive/stream_reader.py

```diff
@@ -0,0 +1,195 @@
+                for new_file in new_files:
+                    # It's possible files and folders are linked up multiple times, this prevents us from getting stuck in a loop
+                    if new_file["id"] in seen:
+                        continue
+                    seen.add(new_file["id"])
+                    file_name = path + self._get_file_name(new_file)
+                    if new_file["mimeType"] == FOLDER_MIME_TYPE:
+                        folder_name = f"{file_name}/"
+                        # check prefix matching in both directions to handle
+                        prefix_matches_folder_name = any(prefix.startswith(folder_name) for prefix in prefixes)
+                        folder_name_matches_prefix = any(folder_name.startswith(prefix) for prefix in prefixes)
+                        if prefix_matches_folder_name or folder_name_matches_prefix or len(prefixes) == 0:
+                            folder_id_queue.append((folder_name, new_file["id"]))
+                        continue
+                    else:
+                        last_modified = datetime.strptime(new_file["modifiedTime"], "%Y-%m-%dT%H:%M:%S.%fZ")
+                        remote_file = GoogleDriveRemoteFile(
+                            uri=file_name, last_modified=last_modified, id=new_file["id"], mimeType=new_file["mimeType"]
+                        )
+                        if self.file_matches_globs(remote_file, globs):
+                            yield remote_file
+                request = service.files().list_next(request, results)
+                if request is None:
+                    break
+
+    def _get_folder_id(self, url):
+        # Regular expression pattern to check the URL structure and extract the ID
+        pattern = r"^https://drive\.google\.com/drive/folders/([a-zA-Z0-9_-]+)$"
+
+        # Find the pattern in the URL
+        match = re.search(pattern, url)
+
+        if match:
+            # The matched group is the ID
+            drive_id = match.group(1)
+            return drive_id
+        else:
+            # If no match is found
+            raise ValueError(f"Could not extract folder ID from {url}")
+
+    def _is_exportable_document(self, mime_type: str):
+        """
+        Returns true if the given file is a Google App document that can be exported.
+        """
+        return mime_type in EXPORTABLE_DOCUMENTS_MIME_TYPES
+
+    def _get_file_name(self, file: Dict[str, Any]):
+        """
+        Returns the file name for the given file.
+
+        Google Docs are exported as Docx to preserve as much formatting as possible, other google app documents are exported as PDFs.
+        For all other files, the file name is returned as is.
+        """
+        if file["mimeType"] == GOOGLE_DOC_MIME_TYPE:
+            return f"{file['name']}.docx"
+        elif self._is_exportable_document(file["mimeType"]):
+            return f"{file['name']}.pdf"
+        else:
+            return file["name"]
```

Comment:
In future, we may also want to customize handling for presentations. Similar to the DOCX format, a PPTX format is probably going to have a lower cost parsing for slide decks, but I don't think we need to prioritize it as of yet and there _might_ be some areas where slides are better parsed as PDFs than as slides. 🤷 (Testing would tell.)

A larger issue, I just realized, is Google Sheets. They don't really fit the 'text documents' mold, and it would be preferably to perhaps ignore them when this parser is selected. (Presumably later on, this connector could  add a "Google Sheets" parser that took the same parse options as Excel and internally converted to Excel before parsing.)

I'm thinking now about other document types like CSVs - whether they would+should get picked up by the text document parser, or if we should ignore them based on their MIME type if the Document Parser is selected.

I wonder if an efficient way to handle this would be to add a MIME type glob pattern. Traditionally, we'd use file extensions for this, but google docs don't have them, and a more robust include/exclude logic would perhaps necessarily use MIME type or some other means of filtering. (Caveat: this is an area I want to understand well and plan for, but I don't know if it should block merging.)

Reply:
You are raising a very good point about robustness of the file detection - what about the following:

There are three "leads" we can leverage:
* The mime type
* The file extension
* The actual file (inspecting the first few bytes) - this works well for docx, pptx and pdf so we should be pretty safe here

I created a PR to add this to the existing unstructured parser in the CDK here: https://github.com/airbytehq/airbyte/pull/31997

Once that is merged I can remove the artificial file extensions from this PR and pass along the mime type (S3 and Azure Blob Storage also have a notion of mime type so it should be possible to extend them in a similar way, but as we have the file inspection it shouldn't be strictly necessary - plus it's always good to have another layer to fall back to as not all objects are guaranteed to have this information).

> A larger issue, I just realized, is Google Sheets. They don't really fit the 'text documents' mold, and it might (?) be preferably to perhaps ignore them when this parser is selected. Presumably later on, this connector could add a "Google Sheets" parser that took the same parse options as Excel and internally converted to Excel before parsing. 🤷

Yeah, I wasn't sure about that either - we already have a dedicated Sheets connector, maybe that's good enough for now. I'm happy to exclude it for the time being and get back to it later.

Reply:
Are you planning to merge https://github.com/airbytehq/airbyte/pull/31997 before this PR is merged? (That seems preferable to me.)

Reply:
Yes, I would like to take that route.